File: runtime/src/main/java/org/pastalab/fray/runtime/TimeDelegate.java
package org.pastalab.fray.runtime;

import java.time.Instant;

public class TimeDelegate {
    public long onNanoTime() {
        return System.nanoTime();
    }

    public long onCurrentTimeMillis() {
        return System.currentTimeMillis();
    }

    public Instant onInstantNow() {
        return Instant.now();
    }

}

File: core/src/main/kotlin/org/pastalab/fray/core/delegates/TimeDelegate.kt
package org.pastalab.fray.core.delegates

import java.time.Instant
import org.pastalab.fray.core.controllers.TimeController
import org.pastalab.fray.runtime.TimeDelegate

class TimeDelegate(val controller: TimeController, val synchronizer: DelegateSynchronizer) :
    TimeDelegate() {
  override fun onNanoTime(): Long =
      synchronizer.runInFrayDoneWithOriginBlockAndNoSkip(
          { controller.nanoTime() },
          { System.nanoTime() },
      )

  override fun onCurrentTimeMillis(): Long =
      synchronizer.runInFrayDoneWithOriginBlockAndNoSkip(
          { controller.currentTimeMillis() },
          { System.currentTimeMillis() },
      )

  override fun onInstantNow(): Instant =
      synchronizer.runInFrayDoneWithOriginBlockAndNoSkip(
          { controller.instantNow() },
          { Instant.now() },
      )
}
